### PR TITLE
coding styles

### PR DESCRIPTION
**Pull request by maelle**

### CODE REVIEW

**Comment by maelle** [comment]
https://twitter.com/lorenzwalthert/status/1494589926251376641

**Comment by maelle** [comment]
@mpadge mentioned https://pypi.org/project/black/ could be mentioned (for Python, not R, but still interesting).

**Review by cderv — COMMENTED** [review]
Nice collection of resources put together and explain. 

I wrote comment in-doc as I was reading. The below could make more sense (or not) after reading those, as I have writting this after commenting in-doc.

Overall, I am thinking about this organization which is close to what you wrote: 

- What is style guide ? 
	- set of rules for the better (clarity, collaboration, ...)
	- Sometime enforce by the language (indentation in python), sometime quite loose like R (where you can add a lot of spaces) and so in the latter more own style exist
- Why does it matters ? 
   - Even alone, it helps easier reading  and upkeep
   - best for collaboration :  proof ? Big organization have that for all language (ex: Google)
   - General resources
- Tools are here to help
   - linters are here for this  (either to show you a problem, or sometime fix it for you)
   - and Most IDE have or can use third-party linters .
   - Usually people who code already apply a default style thanks to IDE (and this is why it is important to use IDE !)
-  Coding Style for  R
   - As said, a lot of them can exist due to the language. 
   - Most known is from organization (tidyverse), but other orgs have their own (mlr)
   - Users can have there own preferences (Mark, Roger, Yihui, ...)
-  Tools to help styling in R
   -  IDE own support (like in RStudio)
   -  Initially, we had linter only: **lintr** - still useful
   -  If you don't mind auto-fixing: **styler** (most generic) **formatR** (opinionated)
   -  Can be setup for R in different tools (VSCODE plugin, EMACS, RStudio addins)
   - Some tools go further with code improvement in context (**pkgcheck**)
   - Coding R chunks in Rmd is also possible ! 
 - Being respectful of coding style when contributing
   - About etiquette and several scenarios
   - One should follow the guide but sometimes maintenaire don't mind doing cosmetic change before merging
- Conclusion

This covers most of the content and my other ideas. Only Markdown styling needs to be place somewhere 😅 

I am just sharing all this so that you can have materials to see how we could adapt (or not) the current content. It seemed easier for now that (re)writing some parts without discussion :)

File: content/post/2022-02-21-code-style/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
```
[the remark this answers is not in the document]
Comment (maelle) [reply content/post/2022-02-21-code-style/index.Rmd:32]: I think neither blogdown nor hugodown have this, or maybe one of them does now :thinking: It's for taking advantages of Hugo figure shortcode. :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (cderv) [reply content/post/2022-02-21-code-style/index.Rmd:32]: No I don't think they have it seems quite useful :) Maybe that worth a Feature Request or a PR in **blogdown**. But I'll remember and maybe I'll do it :)

File: content/post/2022-02-21-code-style/index.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+If you are a contributor to a codebase, you'll probably have to adapt your style to the maintainer's or maintainers' preferences. These preferences might be implicit (seen by reading existing code) or explicit (in a contributing guide). Now, depending on your relation to the maintainers (e.g. is it your first contribution), you might start a discussion about changing some of the conventions, perhaps changing their mind or reaching a compromise (assigning with `=`, but with spaces before and after each equal sign).
+
+If you are the maintainer of a codebase, you'll need to be forgiving when stating your preferences and explaining how to implement them; or you might even want to take matters in your own hands and do some [restyling yourself](https://yihui.org/en/2018/11/cosmetic-changes/).
+
+## Conclusion
+
+In this post we shared documentation of and tooling for coding style. We hope it can help you write or advocate for more readable code. Practice makes perfect, so go forth and participate in code production and reviews. :wink:
+
+[^1]: Note that the Google R style guide inspired the tidyverse styleguide but Google now refers to the tidyverse style guide as R style guide.
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.md:123]:
```suggestion
[^1]: Note that the Google R style guide first inspired the tidyverse styleguide but now Google R style guide is derived from the tidyverse style guide. 
```

File: content/post/2022-02-21-code-style/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+## Resources
+
+One way to develop a sense of coding style is to read a lot of code, but where to learn about coding style explicitly?
+
+### Specific R resources
+
+The most popular, or at least most documented style guide out there is the [tidyverse style guide](https://style.tidyverse.org).
+
+Some variations on it include
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.Rmd:67]: At first I did not catch that, but should we really use the word variation ? It sounds like it is derived and at least some of the rules follow. I see the below list as other different organisation or personal preferences and not specifically a variation of tidyverse one. 

I think we should present it as tidyverse one is the norm and other derived from it. 

```suggestion
Some other style preferences include
```

File: content/post/2022-02-21-code-style/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+-   [Mark Padgham's spacing preferences](https://docs.ropensci.org/pkgcheck/CONTRIBUTING.html#development-guidelines);
+-   [Roger Peng's 8-space indentations](https://simplystatistics.org/posts/2018-07-27-why-i-indent-my-code-8-spaces/);
+-   Yihui Xie's preference for equal sign assignments (as presented in this [issue comment](https://github.com/Robinlovelace/geocompr/issues/319#issuecomment-427376764));
+-   [Google R Style guide](https://google.github.io/styleguide/Rguide.html) with e.g. a preference for BigCamelCase for function names.
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.Rmd:72]:
```suggestion
-   [Google R Style guide](https://google.github.io/styleguide/Rguide.html), a fork of Tidyverse style guide but with variations e.g. a preference for BigCamelCase for function names.
```

File: content/post/2022-02-21-code-style/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+#### lintr
+
+Linting tools will indicate errors and potentially style preference violations.
+
+The [lintr R package by Jim Hester](https://github.com/r-lib/lintr) has a lot of useful linters such as whether code is commented, whether lines are too long etc.
+
+It is pretty common to find *linting tools* in all editors (VSCODE and others) and for most languages.
+To set up lintr with your code editor refer to [its docs](https://github.com/r-lib/lintr#editors-setup).
+Unless you have RStudio IDE and linting is the one included already in the IDE (for errors mainly only, not style preference).
+If you use VSCODE see <https://github.com/REditorSupport/vscode-R/wiki/R-Language-Service#formatting>.
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.Rmd:116]: Formatting in the R extensions uses styler. Maybe we should move the link to part below ? 

Link for VSCODE diagnostic would be https://github.com/REditorSupport/vscode-R/wiki/R-Language-Service#diagnostics

```suggestion
If you use VSCODE, [vscode-R](https://marketplace.visualstudio.com/items?itemName=Ikuyadeu.r) extension includes diagnostic tool - see <https://github.com/REditorSupport/vscode-R/wiki/R-Language-Service#diagnostics>.
```

File: content/post/2022-02-21-code-style/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+To use lintr in testthat see [`expect_lint()`](https://rdrr.io/cran/lintr/man/expect_lint.html).
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.Rmd:118]: I would give a bit more context on this.

```suggestion
Some teams likes to run diagnostic tools as part of their Continuous Integration (CI) workflow. lintr can be used in this context as part of tests suits with testthat - see [`expect_lint()`](https://rdrr.io/cran/lintr/man/expect_lint.html)
```

File: content/post/2022-02-21-code-style/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+Its documentation includes a handy vignette on [customizing styler](https://styler.r-lib.org/articles/customizing_styler.html), for when preferences differ from the default.
+Examples:
+
+-   the [tiny spaceout R package](https://github.com/ropensci-review-tools/spaceout) that adds spaces between code references.
+-   the [grkstyle R package](https://github.com/gadenbuie/grkstyle) holding Garrick Aden-Buie's personal preferences.
+-   the [styler.mlr R package](https://github.com/mlr-org/styler.mlr) implementing the mlr style guide.
+
+The styler package documents some [third-part integration](https://styler.r-lib.org/articles/third-party-integrations.html): you don't have to remember to run styler manually.
+In particular, to use styler on its own CI workflow on GitHub Actions: <https://github.com/r-lib/actions/tree/v2-branch/examples#style-package>
+
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.Rmd:142]: I don't know how you would include a not like box (or maybe this should be in footnote) but I would add a disclaimer on: 

With fixing tools, your original source code will be modified. It is really adviced to use version control with your project - but (place emoji here) who does not use version control when coding nowadays ? If you do not, no worries we got you covered (place link to awesome blog post) 

What do you think ?

File: content/post/2022-02-21-code-style/index.Rmd [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+#### formatr R package
+
+The [formatr R package](https://yihui.org/formatr/) formats R code automatically too, but in a more opinionated way.
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.Rmd:145]: You're right with your question on opiniotated, i would change that

```suggestion
The [formatr R package](https://yihui.org/formatr/) formats R code automatically too, but with less customization possibilities.
```

File: content/post/2022-02-21-code-style/index.Rmd [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+For more tools helping with code improvements, refer to the [R-hub blog post "Workflow automation tools for package developers"](/2020/04/29/maintenance/) including tips on *when* to use such tools.
+
+## A special note on ( R ) Markdown styling
```
Comment (cderv) [inline content/post/2022-02-21-code-style/index.Rmd:149]: Nice way of placing Markdown and R Markdown together ! 😄

**Review by cderv — COMMENTED** [review]
It looks (almost) good to me! 

A few suggested tweak for your to include (or not). 

Thanks for the writing about my first comments!

**Pull request merged**